doc/neps/nep-0027-zero-rank-arrarys.rst

```diff
@@ -0,0 +1,251 @@
+=========================
+NEP 27 — Zero Rank Arrays
+=========================
+
+:Author: Alexander Belopolsky (sasha), transcribed Matt Picus <matti.picus@gmail.com>
+:Status: Draft
+:Type: Informational
+:Created: 2006-06-10
+
+Abstract
+--------
+
+NumPy has both zero rank arrays and scalars. This design document, adapted from
+a `2006 wiki entry`_, describes what zero rank arrays are and why they exist.
+It was transcribed 2018-10-13 into a NEP and links were updated.
+
+Note that some of the information here is dated, for instance indexing of 0-D
+arrays now is now implemented and does not error.
+
+Zero-Rank Arrays
+----------------
+
+Zero-rank arrays are arrays with shape=().  For example:
+
+    >>> x = array(1)
+    >>> x.shape
+    ()
+
+
+Zero-Rank Arrays and Array Scalars
+----------------------------------
+
+Array scalars are similar to zero-rank arrays in many aspects::
+
+
+    >>> int_(1).shape
+    ()
+
+They even print the same::
+
+
+    >>> print int_(1)
+    1
+    >>> print array(1)
+    1
+
+
+However there are some important differences:
+
+* Array scalars are immutable
+* Array scalars have different python type for different data types
+ 
+Motivation for Array Scalars
+----------------------------
+
+Numpy's design decision to provide 0-d arrays and array scalars in addition to
+native python types goes against one of the fundamental python design
+principles that there should be only one obvious way to do it.  In this section
+we will try to explain why it is necessary to have three different ways to
+represent a number.
+
+There were several numpy-discussion threads:
+   
+
+* `rank-0 arrays`_ in a 2002 mailing list thread.
+* Thoughts about zero dimensional arrays vs Python scalars in a `2005 mailing list thread`_]
+
+It has been suggested several times that NumPy just use rank-0 arrays to
+represent scalar quantities in all case.  Pros and cons of converting rank-0
+arrays to scalars were summarized as follows:
+
+- Pros: 
```

Comment:
I wouldn't really suggest changing this doc, but I think this section in particular is out of date.

The main argument I've heard for scalar types is speed -- they are significantly faster than working with 0-d arrays.

The first and third "Pros" here are no longer true. With Python 3, Python uses `operator.index()` for coercing integers and  NumPy scalars can't be relied upon to subclass Python types (e.g., `isinstance(np.int64(1), int)` -> `False`).

Reply:
I find it unlikely that scalars are faster, given that most of our operations start by casting them to 0d arrays. Perhaps the arithmetic has a fast path.

Reply:
so is the conclusion is we should get rid of scalars once we move to python 3?

Reply:
I don't agree with that conclusion - I also think that python 2.7 already uses `__index__`, so if there's a cut-off line here we've already crossed it.

One thing I would like to see is a merge of the scalar types and dtypes - so that `isinstance(np.dtype, type)` is true, and `isinstance(np.float64, np.dtype)` is also true. But that's blocked by #11998 right now.

Reply:
Scalars absolutely have a fast path.  It is in numpy/numpy/core/src/umath/scalarmath.c.src among other places.

Reply:
@eric-wieser - Scalars are indeed fast-tracked for arithmetic (`scalarmath.c.src`):
```
In [3]: a = np.array(1.)

In [4]: %timeit a * a
1000000 loops, best of 3: 541 ns per loop

In [5]: a = np.float64(1.)

In [6]: %timeit a * a
10000000 loops, best of 3: 91.6 ns per loop
```